wrappers/hnygingonic/gingonic.go

```diff
@@ -0,0 +1,76 @@
+package hnygingonic
+
+import (
+	"context"
+
+	"github.com/gin-gonic/gin"
+	"github.com/honeycombio/beeline-go"
+	"github.com/honeycombio/beeline-go/trace"
+	"github.com/honeycombio/beeline-go/wrappers/common"
+)
+
+// Middleware wraps httprouter handlers. Since it wraps handlers with explicit
+// parameters, it can add those values to the event it generates.
+func Middleware() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		// get a new context with our trace from the request, and add common fields
+		ctx, span := common.StartSpanOrTraceFromHTTP(c.Request)
+		defer span.Send()
+		// Add the span context to the gin context as we need to be able to pass
+		// this context around our gin application
+		c.Set("beeline-middleware-context", ctx)
+		// push the context with our trace and span on to the request
+		c.Request = c.Request.WithContext(ctx)
+
+		// replace the writer with our wrapper to catch the status code
+		wrappedWriter := common.NewResponseWriter(c.Writer)
+
+		// pull out any variables in the URL, add the thing we're matching, etc.
+		for _, param := range c.Params {
+			span.AddField("handler.vars."+param.Key, param.Value)
+		}
+
+		// pull out any GET query params
+		for key, value := range c.Request.URL.Query() {
+			if len(value) > 1 {
+				span.AddField("handler.query."+key, value)
+			} else if len(value) == 1 {
+				span.AddField("handler.query."+key, value[0])
+			} else {
+				span.AddField("handler.query."+key, nil)
+			}
+		}
```

Comment:
This is tempting and dangerous for any publicly accessible web services. All the spew on the internet that attempts to hit your webserver with paths like ../../../../../../password.txt or whatevs also often throws garbage query parameters in your URLs. By automatically accepting all query parameters as fields you are ceding control of your dataset schema to the public internet - a recipe for disaster. I say this as one who has fallen for this trap at least 3 times. A far more stable approach is to explicitly add only the parameters that you care about when you start to process them and accept that it does require a little more work throughout the rest of your codebase.

There's one exception to this advice, and I don't know whether gin does this - but it's the reason this is down on this block intsead of above on the `c.Params` block. If the contents of `c.Params` does not come from the list of query parameters in the URL but instead is registered with gin as part of the handler creation, then this is totally fine since they've effectively already been explicitly declared as desired arguments. However, I'm pretty sure this block is grabbing things from the URL itself, so is probably not safe.

Reply:
Ah, that totally makes sense, I'll update that. My initial thoughts on it would be to add a param to the `Middleware` function that took the GET params you expect to deal with in your requests.

e.g.

```go
func Middleware(getParams map[string]struct{}) gin.HandlerFunc {
	return func(c *gin.Context) {
		...
		// pull out any GET query params
		if getParams != nil {
			for key, value := range c.Request.URL.Query() {
				if _, ok := getParams[key]; ok {
					if len(value) > 1 {
						span.AddField("handler.query."+key, value)
					} else if len(value) == 1 {
						span.AddField("handler.query."+key, value[0])
					} else {
						span.AddField("handler.query."+key, nil)
					}
				}
			}
		}
		...
	}
}
```

As for the `c.Params` I believe that is setup from the url string defined in the router

e.g.

```go
func main() {
	...
	router.GET("/person/:name/pets", listPets) // In this route :name is a param
	...
}
```